src/test/BasicTests.hs

```diff
@@ -0,0 +1,139 @@
+-- | This module contains basic tests for 'haskell-source-transformations'
+module BasicTests
+  ( basicTests
+  )
+where
+
+import           Test.Hspec                     ( Spec
+                                                , Expectation
+                                                , describe
+                                                , it
+                                                , shouldBe
+                                                )
+import           Test.HUnit.Base                ( assertFailure )
+
+import           Control.Monad                  ( void )
+
+import           Language.Haskell.Exts          ( Module
+                                                , ParseResult(..)
+                                                , parseModule
+                                                )
+import           Language.Haskell.Exts.Pretty   ( Pretty
+                                                , prettyPrint
+                                                )
+
+import           HST.Application                ( processModule
+                                                , specialCons
+                                                )
+import           HST.Environment.FreshVars      ( PMState(..)
+                                                , evalPM
+                                                )
+
+
+basicTests :: Spec
+basicTests = describe "Basic unit tests" tests
+
+-- | Parses a given string to a module and fails if parsing is not
+--   successful.
+parseTestModule :: String -> IO (Module ())
+parseTestModule modStr = case parseModule modStr of
+  ParseOk modul        -> return $ void modul
+  ParseFailed _ errMsg -> assertFailure errMsg
+
+-- | The default state used for pattern matching.
+defaultState :: PMState
+defaultState = PMState { nextId      = 0
+                       , constrMap   = specialCons
+                       , matchedPat  = []
+                       , trivialCC   = False
+                       , opt         = True
+                       , debugOutput = ""
+                       }
+
+-- | Test cases for 'processModule'.
+tests :: Spec
+tests = do
+  it "should accept a simple function" $ do
+    mod1 <- parseTestModule "module A where\nf :: a -> a\nf x = x"
+    let mod2 = evalPM (processModule mod1) defaultState
+    mod2 `prettyShouldBe` mod1
```

Comment:
This change should not be applied now, but this is my suggestion on how to structure the tests once both HSE and `ghc-lib` are implemented as available frontends:
```suggestion
  let simpleFuncTest = "module A where\nf :: a -> a\nf x = x"
  it "should accept a simple function (HSE)" $
    testHSE simpleFuncTest simpleFuncTest
  it "should accept a simple function (ghc-lib)" $
    testGHC simpleFuncTest simpleFuncTest
```
This doesn't increase the size of the tests by much. Of course, `testHSE` and `testGHC` have to be added for this. They would both take two strings, parse them with HSE or `ghc-lib`, apply the pattern matching compiler to one of them, print them and compare the results.

Reply:
Ideally, there would be tests for the conversion from HSE and GHC-lib to our syntax and all other tests use our syntax only. This way we don't have to run tests twice. However, for a transitional period, the approach using `testHSE` and `testGHC` should work fine.